include/aws/common/math.h

```diff
@@ -0,0 +1,189 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/**
+ * Multiplies a * b. If the result overflows, returns 2^64 - 1.
+ */
+static inline uint64_t aws_common_mul_u64_saturating(uint64_t a, uint64_t b) {
+#if defined(__x86_64__) && defined(__GNUC__)
+    // We can use inline assembly to do this efficiently on x86-64 and x86.
+
+    // we specify rdx as an output, rather than a clobber, because we want to
+    // allow it to be allocated as an input register
+
+    uint64_t rdx;
+    __asm__(
+        "mulq %q2\n" // rax * b, result is in RDX:RAX, OF=CF=(RDX != 0)
+        "cmovc %q3, %%rax\n"
+        : /* in/out: %rax = a, out: rdx (ignored) */ "+a" (a), "=d" (rdx)
+        : /* in: anything (imm/reg/memory) */ "g" (b),
+          /* in: saturation value (reg/memory) */ "rm" (~0LL)
+        : /* clobbers: cc */ "cc"
+    );
+    (void)rdx; // suppress unused warnings
+    return a;
+#else
+    uint64_t x = a * b;
+    if (a != 0 && (a > 0xFFFFFFFF || b > 0xFFFFFFFF) && x / a != b) {
+        return ~(uint64_t)0;
+    } else {
+        return x;
+    }
+#endif
+}
+
+/**
+ * Multiplies a * b and returns the truncated result in *r. If the result
+ * overflows, returns 0, else returns 1.
+ */
+static inline int aws_common_mul_u64_checked(uint64_t a, uint64_t b, uint64_t *r) {
+#if defined(__x86_64__) && defined(__GNUC__)
+    // We can use inline assembly to do this efficiently on x86-64 and x86.
+
+    int flag;
+    uint64_t result;
+    __asm__(
+        "mulq %q3\n" // rax * b, result is in RDX:RAX, OF=CF=(RDX != 0)
+        "setno %b1\n" // rdx = (OF = 0)
+        "and $0xFF, %1\n" // mask out flag
+        : /* out: %rax */ "=a" (result), "=d" (flag)
+        : /* in: reg or memory (possibly rax/rdx), imm/reg/memory for 2nd operand */
+          "a" (a), "g" (b)
+        : /* clobbers: cc */ "cc"
+    );
+    *r = result;
+    return flag;
+#else
+    uint64_t x = a * b;
+    *r = x;
+    if (a != 0 && (a > 0xFFFFFFFF || b > 0xFFFFFFFF) && x / a != b) {
+        return 0;
+    } else {
+        return 1;
+    }
+#endif
+}
+
+/**
+ * Multiplies a * b. If the result overflows, returns 2^32 - 1.
+ */
+static inline uint32_t aws_common_mul_u32_saturating(uint32_t a, uint32_t b) {
+#if (defined(__i386__) || defined(__x86_64__)) && defined(__GNUC__)
+    // We can use inline assembly to do this efficiently on x86-64 and x86.
+
+    // we specify edx as an output, rather than a clobber, because we want to allow it
+    // to be allocated as an input register
+    uint32_t edx;
+    __asm__(
```

Comment:
For compatibility should these be just asm instead of __asm__?

Reply:
Same comments as above for this.... just transpose them for 32 bit arithmetic.

Reply:
Nope. The difference is `asm` will result in compile failures in a strictly-conforming mode in GCC (e.g. `--std=c99 --pedantic`).